### DISCUSSION TITLE
Regression in Gradle version catalog handling (disable major updates)?

### DISCUSSION THREAD

**Discussion by szpak** [renovatebot/renovate#27303] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

In [Spock](https://github.com/spockframework/spock/) we use both mockito4 and mockito5 in different modules for regression testing with both the versions. We have in configured as [follows](https://github.com/spockframework/spock/blob/789f66aacfce0c8be22af90fae73690b2b62bc0c/gradle/libs.versions.toml#L8-L9) using the Gradle version catalog:

```
[versions]
mockito4 = '4.11.0'
mockito5 = '5.10.0'
...

[libraries]
mockito4 = { module = "org.mockito:mockito-core", version.ref = "mockito4" }
mockito5 = { module = "org.mockito:mockito-core", version.ref = "mockito5" }
...
```

To keep mockito4 as 4.x we [added](https://github.com/spockframework/spock/blob/789f66aacfce0c8be22af90fae73690b2b62bc0c/.github/renovate.json5#L14-L17):
```
  "packageRules": [
    // Disable major version upgrades for Groovy versions
    // as we want to keep one for each major version.
    // The same applies for mockito as we test with both 4 and 5
    {
      "matchPackagePrefixes": ["org.codehaus.groovy:", "org.apache.groovy:", "org.mockito:"],
      "matchManagers": ["regex"],
      "matchUpdateTypes": ["major"],
      "enabled": false
    }
  ]
```

It worked fine, but suddenly, on 24 Jan 2024, the [following PR](https://github.com/spockframework/spock/pull/1871/files) was generated, bumping mockito4 to 5.x:
```
- mockito4 = '4.11.0'
+ mockito4 = '5.10.0'
  mockito5 = '5.10.0'
```

It seems there is some regression to have it not working anymore. I took a look at the changes around that time, but I haven't found a obvious culprit :-/. Only https://github.com/renovatebot/renovate/pull/26815/ is somehow related to regexps.

Here's a [minimal reproducer](https://github.com/szpak-forks/renovate-repro-mockito5-spock/pull/2/files). Do you know how bring the old functionality back?

### Relevant debug logs

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by Churro** [comment, marked as answer]
Your package rule has no effect, as it is limited to dependencies matched by the `regex` manager. Native Gradle version catalog handling is done by the `gradle` manager. It also doesn't distinguish between the two occurrences of mockito. You probably only noticed this on January 24th, because mockito 5.10.0 was released on that day.

In your case, to keep `mockito4` as-is but continue to receive renovate updates for `mockito5`, you could use the following:
```json
  "packageRules": [
    {
      "matchPackagePrefixes": ["org.mockito:"],
      "matchCurrentValue": "/^4\\./",
      "allowedVersions": "(,5.0)"
    }
  ]
```

**Reply by szpak** [reply]
Thanks, it helped.
